core/src/main/java/hivemall/anomaly/ChangeFinderUDF.java

```diff
@@ -0,0 +1,540 @@
+                yScoreHistory.addFirst(yScore);
+                yScoreRunningSum -= yScoreHistory.getLast();
+                yScoreHistory.removeLast();
+                yScore = yScoreRunningSum / yRunningWindowSize;
+            }
+            yTrain(xScore);
+        }
+
+        callCount++;
+        Object[] output = new Object[4];
+        output[0] = new DoubleWritable(xScore);
+        output[1] = new BooleanWritable(xScore >= xThreshold);
+        output[2] = new DoubleWritable(yScore);
+        output[3] = new BooleanWritable(yScore >= yThreshold);
+        return output;
+    }
+
+    private void xTrain() {
+        //update mean vector
+        //\hat{µ} := (1-r) \hat{µ} + r x
+        xMeanEstimate = xMeanEstimate.mapMultiplyToSelf((1.d - xForgetfulness))
+                                     .add(x.mapMultiply(xForgetfulness));
+
+        //residuals (x - \hat{µ})
+        RealVector xResidual0 = x.copy().subtract(xMeanEstimate);
+        RealVector[] xResiduals = new RealVector[xRunningWindowSize];
+        for (int i = 0; i < xRunningWindowSize; i++) {
+            xResiduals[i] = xHistory.get(i).subtract(xMeanEstimate);
+        }
+
+        //update covariance matrices
+        //C_j := (1-r) C_j + r (x_t - \hat{µ}) (x_{t-j} - \hat{µ})'
+        xCovar0 = xCovar0.scalarMultiply(1.d - xForgetfulness)
+                         .add(xResidual0.mapMultiply(xForgetfulness).outerProduct(xResidual0));
+        for (int i = 0; i < (xHistoryFull ? xRunningWindowSize : callCount); i++) {
+            xCovar[i] = xCovar[i].scalarMultiply(1.d - xForgetfulness).add(
+                xResidual0.mapMultiply(xForgetfulness).outerProduct(xResiduals[i]));
+        }
+
+        if (xCovarFull) {
+            //solve C_j = ∑_{i=1}^{k} A_i C_{j-i} where j = 1..k, C_{-i} = C'_i, k = xRunningWindowSize for A_i
+            //combine RHS covariance matrices into one large symmetric matrix
+            /*
+             *  xCovarsCombined (symmetric, invertible, k*dimensions by k*dimensions)
+             *
+             * /C_0     |C_1     |C_2     | .  .  .   |C_{k-1} \
+             * |--------+--------+--------+           +--------|
+             * |C_1'    |C_0     |C_1     |               .    |
+             * |--------+--------+--------+               .    |
+             * |C_2'    |C_1'    |C_0     |               .    |
+             * |--------+--------+--------+                    |
+             * |   .                         .                 |
+             * |   .                            .              |
+             * |   .                               .           |
+             * |--------+                              +-------|
+             * \C_{k-1}'| .  .  .                      |C_0    /
+             */
+            RealMatrix[][] xCovarsCombinedRaw =
+                    new RealMatrix[xRunningWindowSize][xRunningWindowSize];
+            for (int i = 0; i < xRunningWindowSize; i++) {
```

Comment:
@L3Sota It seems Toeplitz matrix initialization has a bug.
https://github.com/myui/hivemall/blob/feature/changefinder1d/core/src/main/java/hivemall/utils/math/MatrixUtils.java#L118